### ISSUE TITLE
Can we delete FullSolveMultiApp on initial after it finishes its job?

### ISSUE THREAD

**Issue by YaqiWang** [idaholab/moose#7767]
### Description of the enhancement or error report
We have cases that have mutlple FullSolveMultiApp running on initial. We'd like to delete them once they are executed and the transfer are done in order to save memory.

These probably requires rewriting `FEProblem::execMultiApps()`, which currently have three stages: transfer to multiapp, multiapp solve, transfer from multiapp. We will need to reorder them with a loop through multiapps, and perform the transfers related with each multiapp. We then decide if the multiapp should deleted with a certain rule, for example, it is executed on initial only.

### Rationale for the enhancement or information for reproducing the error
We do want to reduce the memory usage for the large realistic simulations.

### Identified impact
(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)
New enhanced feature.

**Comment by friedmud** [comment]
Both of these things sounds reasonable to me.  A couple of notes on this:

1.  Deleting after execution should be an _option_... even for `initial` multiapps.  It could be the case that you want to run a full solve and then interpolate from it each timestep or something.

2.  I can't think of anything _bad_ that would happen from reordering the `execMultiApps()` loops like you say.  In fact... it would help with one particular thing: it would allow the resolution of dependencies between multiple MultiApps executed during the same `execute_on`.

**Comment by YaqiWang** [comment]
MOOSE team, Hans Hammer at TAMU wants to run a large simulation with `FullSolveMultiApp` for some initial setup. We really need this issue resolved because the multiapp uses ridiculous amount of memory and not need to keep them around after the transfer. Can we get the low-hanging fruit quickly.

**Comment by permcody** [comment]
Tag @idaholab/moose-team

**Comment by aeslaughter** [comment]
No, I have a version of FullSolveMultiapp that can re-solve over and over again, a complete problem. This is how I do the levelset reinitialization step and it probably has other uses.

**Comment by YaqiWang** [comment]
This is optional controlled by input.

**Comment by fdkong** [comment]
It would be a good option for users, but do not know how challenging to add.

**Comment by permcody** [comment]
Should be fairly easy to add. @fdkong if you want to work on this, that'd be awesome. 

I recommend adding a "deleteObject" method to the main warehouse BASE CLASS. Do **not** make that method templated, instead just delete any object by a given name in the warehouse for simplicity. We'll also want to cleanup the associated InputParameters object that sits in the InputParameters warehouse too when an object is deleted from the warehouse. That's it!

There's always a little bit more before this is completely robust but we don't have to worry about this part. In general, if an object is deleted it could break the whole simulation. Imagine deleting the only kernel active on a particular variable, or removing a material supplying properties to other objects. In a perfect world, we'd run integrity checks every time the system state changes but we already aren't doing that with the Control system so I wouldn't worry about it here.

**Comment by aeslaughter** [comment]
This will probably need to step from FEProblemBase so that it can go into
the MultiAppWarehouse and the InputParameterWarehouse. You will need to add
a similar delete method to the InputParameterWarehouse, since it does not
inherit from the MooseObjectWarehouse. Also, you may need to implement
delete methods up the chain of objects to clean up all the different shared
pointers, there are nested lists in the warehouses. None of this is very
difficult, just making you aware.

**Comment by YaqiWang** [comment]
Thanks guys! I am sure this is not only a new capability but also can make MOOSE better.

**Comment by permcody** [comment]
@YaqiWang - Do we need to work on this anytime soon?

**Comment by YaqiWang** [comment]
Not very because we currently implemented a new capability of restarting transport system from a previous solve. But it should be implemented in the future although it may take us some time.
